### PR TITLE
feat(notify): add option to show duration

### PR DESCRIPTION
**Pull request by RrybaN**
**lib.notify New Option**
```typescript
showDuration?: boolean -- Option to show duration bar
```

**Example:**
```lua
lib.notify({
    title = 'Custom',
    description = 'Custom',
    showDuration = true,
    icon = 'eye',
    iconColor = '#C53030'
})
```

**Preview:**

Screenshot 2024-04-08 231022

### CODE REVIEW

**Comment by Demigod916** [comment]
hot

**Comment by FjamZoo** [comment]
Good looks

**Review by LukeWasTakenn — COMMENTED** [review]
Code looks good, though there is this interesting issue of not using a mantine color so if you  use a custom hex color and want to display the notification progress it just looks weird.

image

```lua
  debugData<NotificationProps>([
    {
      action: 'notify',
      data: {
        title: 'Success',
        description: 'Notification description',
        type: 'success',
        id: 'pogchamp',
        duration: 200000,
        showDuration: true,
        iconColor: '#d5eb31',
      },
    },
  ]);
  debugData<NotificationProps>([
    {
      action: 'notify',
      data: {
        title: 'Success',
        description: 'Notification description',
        type: 'success',
        duration: 200000,
        showDuration: false,
        iconColor: '#d5eb31',
      },
    },
  ]);
```

Not entirely sure what we can do about this, maybe perhaps what mantine is doing and just set the background color of the avatar to the same color and lover the alpha?

I'm open to suggestions.

**Comment by RrybaN** [comment]
I think the problem here is that the mantine version is not on the latest version, the latest version can use rgba or hex on `Avatar` component. But either way i found a another solution for this, instead of using `Avatar` component i used `ThemeIcon` for this.

**Preview:**

Screenshot 2024-04-11 191246

File: web/src/features/notifications/NotificationWrapper.tsx [hunk 1]
```diff
@@ -103,11 +103,24 @@ const exitAnimationBottom = keyframes({
...
 });
 
+const durationCircle = keyframes({
+  '0%': { strokeDasharray: `0, ${15.1 * 2 * Math.PI}` },
+  '100%': { strokeDasharray: `${15.1 * 2 * Math.PI}, 0` },
+});
+
+const isValidHexColor = (hex: string) => {
+  return /^#[0-9A-F]{6}$/i.test(hex)
+}
```
Comment (LukeWasTakenn) [inline web/src/features/notifications/NotificationWrapper.tsx:113]: Hex or mantine color aren't the only options, you can pass colors as rgb or hsl like such: `rgb(255, 255, 110)` which results in the icon not being colored

image

**Review by LukeWasTakenn — CHANGES_REQUESTED** [review]
Looks great overall though I can also see someone not wanting to show the progress but wanting the light background color of ThemeIcon and I'm not sure what would be the best way to handle that 🤔 

Since the size of the icon is now a bit different it looks weird if you have a notification with an icon background and another one without it.

image


Maybe just constantly show the background color with ThemeIcon? Though I can also see some people *not* preferring this.

**Comment by RrybaN** [comment]
Noted, but which one do you prefer the light background or constantly show the background color? Or do you have anything else in your mind.

File: web/src/features/notifications/NotificationWrapper.tsx (hunk 1 above)
Comment (RrybaN) [reply web/src/features/notifications/NotificationWrapper.tsx:113]: Oh yes, i clearly forgot about this

**Comment by LukeWasTakenn** [comment]
> Noted, but which one do you prefer the light background or constantly show the background color? Or do you have anything else in your mind.

The light background color is good the way it is now, just have it always present even when using `iconColor` that way we avoid the difference in spacing, and also need to make it work with different color formats.

I think we also make it display duration by default and make it an optional opt out by setting `showDuration: false`.

**Comment by RrybaN** [comment]
> and also need to make it work with different color formats.

For this one can i add a package `tinycolor2`
https://www.npmjs.com/package/tinycolor2

so we can just convert all available color formats to rgba, because the current mantine version only accepts hex, rgb, rgba

**Comment by RrybaN** [comment]
For this one show duration is defaulted to show and also background color is always present even when using `iconColor`. Can now use `'hex' | 'hexa' | 'rgb' | 'rgba' | 'hsl' | 'hsla'` for `iconColor`.

**Preview:**

Screenshot 2024-04-12 040806

**Review by thelindat — APPROVED** [review]
LGTM, just need to add showDuration to the Lua and TS NotifyProps type.

**Review by LukeWasTakenn — APPROVED** [review]
Looks great, thanks for the PR!

Would you also mind updating the docs with the `showDuration` prop?

**Comment by RrybaN** [comment]
Sure, I don't mind updating the docs. I'm glad to help with anything.

**Pull request merged**